src/core/server/integration_tests/saved_objects/migrations/zero_downtime/rerun_same_version.test.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Path from 'path';
+import fs from 'fs/promises';
+import { createTestServers, type TestElasticsearchUtils } from '@kbn/core-test-helpers-kbn-server';
+import '../jest_matchers';
+import { getKibanaMigratorTestKit } from '../kibana_migrator_test_kit';
+import { delay, parseLogFile } from '../test_utils';
+import { getBaseMigratorParams, getFooType, getBarType } from './base.fixtures';
+
+export const logFilePath = Path.join(__dirname, 'rerun_same_version.test.log');
+
+describe('ZDT upgrades - rerun migration on same version', () => {
+  let esServer: TestElasticsearchUtils['es'];
+
+  const startElasticsearch = async () => {
+    const { startES } = createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'basic',
+        },
+      },
+    });
+    return await startES();
+  };
+
+  beforeAll(async () => {
+    await fs.unlink(logFilePath).catch(() => {});
+    esServer = await startElasticsearch();
+  });
+
+  afterAll(async () => {
+    await esServer?.stop();
+    await delay(10);
+  });
+
+  const createBaseline = async () => {
+    const fooType = getFooType();
+    const barType = getBarType();
+    const { runMigrations } = await getKibanaMigratorTestKit({
+      ...getBaseMigratorParams(),
+      types: [fooType, barType],
+    });
+    await runMigrations();
+  };
+
+  it('should perform a no-op upgrade', async () => {
```

Comment:
To check my understanding: does the ZDT algo actually re-run queries to check for outdated documents?

Reply:
It currently does. it's probably not necessary if we were to check the `docVersions` property, but I didn't want to optimize too soon (and checking for docs everytime is safest anyway)